elyra/contents/file_parser.py

```diff
@@ -0,0 +1,279 @@
+        pass
+
+    @abstractmethod
+    def get_next_code_chunk(self, filepath):
+        # Implements a generator for lines of code for the specified operation type
+        pass
+
+    @abstractmethod
+    def get_resources(self, filepath):
+        # Returns a model dictionary of all the regex matches for each key in the regex dictionary
+        pass
+
+
+class NotebookFileParser(FileParser):
+
+    _type = "notebook"
+
+    def __init__(self):
+        # super().__init__(operation_filepath)
+        # self.language = None
+        self.parser = None
+
+    def type(self):
+        return self._type
+
+    def file_as_list_of_code_chunks(self, filepath) -> List[str]:
+        code_chunks = []
+        with open(filepath) as f:
+            self.notebook = nbformat.read(f, as_version=4)
+            language = self.notebook['metadata']['kernelspec']['language']
+
+            if language == 'python':
+                self.parser = PythonScriptParser()
+            elif language == 'r':
+                self.parser = RScriptParser()
+
+            for cell in self.notebook.cells:
+                if cell.source and cell.cell_type == "code":
+                    code_chunks.append(cell.source.split('\n'))
+        return [item for subchunk in code_chunks for item in subchunk]
+
+    def get_next_code_chunk(self, filepath):
+        with open(filepath) as f:
+            self.notebook = nbformat.read(f, as_version=4)
+            language = self.notebook['metadata']['kernelspec']['language']
+
+            if language == 'python':
+                self.parser = PythonScriptParser()
+            elif language == 'r':
+                self.parser = RScriptParser()
+
+            for cell in self.notebook.cells:
+                if cell.source and cell.cell_type == "code":
+                    yield cell.source.split('\n')
+
+    def get_resources(self, filepath):
+        model = dict()
+        model["env_list"] = dict()
+        model["inputs"] = dict()
+        model["outputs"] = dict()
+
+        for chunk in self.get_next_code_chunk(filepath):
+            if chunk:
+                # Notebook cells must be further split into individual lines
+                for line in chunk:
+                    matches = self.parser.parse_environment_variables(line)
+                    for key, match in matches:
+                        model[key][match.group(1)] = match.group(2)
+
+        return model
+
+
+class PythonFileParser(FileParser):
+
+    _type = "python"
+
+    '''
+    def __init__(self):
+        # super().__init__(operation_filepath)
+    '''
+
+    def type(self):
+        return self._type
+
+    def file_as_list_of_code_chunks(self, filepath) -> List[str]:
+        with open(filepath) as f:
+            code_chunks = f.readlines()
+        return [c.strip() for c in code_chunks]
+
+    def get_next_code_chunk(self, filepath):
+        with open(filepath) as f:
+            for line in f:
```

Comment:
Does this work?  I would have figured you'd need to call `readlines()` somewhere first.

Once settled we could move this into `FileParser` so RFileParser could use it also as the "default" implementation with `NotebookParser` overriding with its own implementation.

Reply:
It does! I was also surprised. I haven't tested with a very large file yet but it kept showing up as the recommended way in my search. 

> Once settled we could move this into FileParser so RFileParser could use it also as the "default" implementation with NotebookParser overriding with its own implementation.
Makes sense!